scaffoldmaker/utils/tubemesh.py

```diff
@@ -243,6 +243,107 @@ def generatetubemesh(region,
                 result = element.setNodesByIdentifier(eft, nodeIdentifiers)
                 elementIdentifier = elementIdentifier + 1
 
+    # Create texture coordinates
+    textureCoordinates = getOrCreateTextureCoordinateField(fm)
+    textureNodetemplate1 = nodes.createNodetemplate()
+    textureNodetemplate1.defineField(textureCoordinates)
+    textureNodetemplate1.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_VALUE, 1)
+    textureNodetemplate1.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D_DS1, 1)
+    textureNodetemplate1.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D_DS2, 1)
+    if useCrossDerivatives:
+        textureNodetemplate1.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D2_DS1DS2, 1)
+
+    textureNodetemplate2 = nodes.createNodetemplate()
+    textureNodetemplate2.defineField(textureCoordinates)
+    textureNodetemplate2.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_VALUE, 2)
+    textureNodetemplate2.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D_DS1, 2)
+    textureNodetemplate2.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D_DS2, 2)
+    if useCrossDerivatives:
+        textureNodetemplate2.setValueNumberOfVersions(textureCoordinates, -1, Node.VALUE_LABEL_D2_DS1DS2, 2)
+
+    bicubichermitelinear = eftfactory_bicubichermitelinear(mesh, useCrossDerivatives)
+    eftTexture = bicubichermitelinear.createEftBasic()
+
+    elementtemplate1 = mesh.createElementtemplate()
+    elementtemplate1.setElementShapeType(Element.SHAPE_TYPE_CUBE)
+    elementtemplate1.defineField(textureCoordinates, -1, eftTexture)
+
+    eftTexture2 = bicubichermitelinear.createEftOpenTube()
+    elementtemplate2 = mesh.createElementtemplate()
+    elementtemplate2.setElementShapeType(Element.SHAPE_TYPE_CUBE)
+    elementtemplate2.defineField(textureCoordinates, -1, eftTexture2)
+
+    # Calculate texture coordinates and derivatives
+    uTexture = []
+    d1Texture = []
+    d2Texture = []
+    for n3 in range(elementsCountThroughWall + 1):
+        for n2 in range(elementsCountAlong + 1):
+            for n1 in range(elementsCountAround + 1):
+                u = [ 1.0 / elementsCountAround * n1,
+                    1.0 / elementsCountAlong * n2,
+                    1.0 / elementsCountThroughWall * n3]
+                d1 = [1.0 / elementsCountAround, 0.0, 0.0]
+                d2 = [0.0, 1.0 / elementsCountAlong, 0.0]
+                uTexture.append(u)
+                d1Texture.append(d1)
+                d2Texture.append(d2)
+
+    nodeIdentifier = nextNodeIdentifier
+    for n in range(len(uTexture)):
+        if n%(elementsCountAround+1) == 0.0:
```

Comment:
== 0, not a floating point.
Also, I like spaces around operators % and +.
That said, I would loop as follows:
```
for n3 in range(elementsCountThroughWall + 1):
    for n2 in range(elementsCountAlong + 1):
        for n1 in range(elementsCountAround):
```
... and then compute coordinates from those precalculated around one slice.